pyqtorch/parametric.py

```diff
@@ -0,0 +1,271 @@
+        jC[unitary_idx:, unitary_idx:, :] = jU
+        return jC
+
+
+class CRX(ControlledRotationGate):
+    def __init__(
+        self,
+        control: int | list[int],
+        target: int,
+        param_name: str,
+        apply_fn_type: ApplyFn = DEFAULT_APPLY_FN,
+    ):
+        super().__init__("X", control, target, param_name, apply_fn_type)
+
+
+class CRY(ControlledRotationGate):
+    def __init__(
+        self,
+        control: int | list[int],
+        target: int,
+        param_name: str = "",
+        apply_fn_type: ApplyFn = DEFAULT_APPLY_FN,
+    ):
+        super().__init__("Y", control, target, param_name, apply_fn_type)
+
+
+class CRZ(ControlledRotationGate):
+    def __init__(
+        self,
+        control: list[int],
+        target: int,
+        param_name: str = "",
+        apply_fn_type: ApplyFn = DEFAULT_APPLY_FN,
+    ):
+        super().__init__("Z", control, target, param_name, apply_fn_type)
+
+
+class CPHASE(Parametric):
+    n_params = 1
+
+    def __init__(
+        self,
+        control: int | list[int],
+        target: int,
+        param_name: str,
+        apply_fn_type: ApplyFn = DEFAULT_APPLY_FN,
+    ):
+        super().__init__("S", target, param_name, apply_fn_type)
+        self.register_buffer("identity", torch.eye(2**self.n_qubits, dtype=DEFAULT_MATRIX_DTYPE))
+        self.control = [control]
+        self.apply_fn = APPLY_FN_DICT[apply_fn_type]
+
+    def matrices(self, thetas: torch.Tensor) -> torch.Tensor:
+        theta = thetas.squeeze(0) if thetas.ndim == 2 else thetas
+        batch_size = len(theta)
+        mat = self.identity.repeat((batch_size, 1, 1))
+        mat = torch.permute(mat, (1, 2, 0))
+        phase_rotation_angles = torch.exp(torch.tensor(1j) * theta).unsqueeze(0).unsqueeze(1)
+        mat[-1, -1, :] = phase_rotation_angles
+        return mat
+
+    def apply(self, matrices: torch.Tensor, state: torch.Tensor) -> torch.Tensor:
+        batch_size = matrices.size(-1)
+        return self.apply_fn(state, matrices, self.qubit_support, self.n_qubits, batch_size)
+
+    def forward(self, state: torch.Tensor, thetas: torch.Tensor) -> torch.Tensor:
+        mats = self.matrices(thetas)
+        return self.apply_operator(mats, state)
+
+
+class U(Parametric):
+    n_params = 3
+
+    def __init__(
+        self, target: int, param_names: list[str], apply_fn_type: ApplyFn = DEFAULT_APPLY_FN
+    ):
+        """
+        Represents a parametrized arbitrary rotation along the axes of the Bloch sphere.
+
+        The angles `phi, theta, omega` in tensor format, applied as:
+
+            U(phi, theta, omega) = RZ(omega)RY(theta)RZ(phi)
+
+        Arguments:
+            qubits (ArrayLike): The target qubits for the U gate. It should be a list of qubits.
+            n_qubits (int): The total number of qubits in the circuit.
+
+        """
+        self.param_names = param_names
+        super().__init__("X", target, param_name="")
+
+        self.register_buffer(
```

Comment:
It feels these projector matrices might be named a bit more meaningfully.

Reply:
fully agree, not my creation though. i dont think i can do a refactor of the full library (including analog for example). i will create issues and assign it to people

Reply:
Well, this PR is a refactoring of the whole library 🤣 But yes, I agree on opening up new issues here.